src/utilities/xml/XMLValidator.hpp

```diff
@@ -0,0 +1,125 @@
+*  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER(S), ANY CONTRIBUTORS, THE UNITED STATES GOVERNMENT, OR THE UNITED
+*  STATES DEPARTMENT OF ENERGY, NOR ANY OF THEIR EMPLOYEES, BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+*  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+*  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+*  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***********************************************************************************************************************/
+
+#ifndef UTILITIES_XML_XMLVALIDATOR_HPP
+#define UTILITIES_XML_XMLVALIDATOR_HPP
+
+#include "../UtilitiesAPI.hpp"
+
+#include "../core/Path.hpp"
+#include "../core/Logger.hpp"
+#include "../core/LogMessage.hpp"
+#include "../core/Optional.hpp"
+#include "../core/StringStreamLogSink.hpp"
+
+#include <boost/optional.hpp>
+
+#include <string>
+
+namespace openstudio {
+
+enum class XMLValidatorType
+{
+  XSD,
+  XSLTSchematron,
+  Schematron
+};
+
+namespace detail {
+  struct XMLInitializer;
+}
+
+class UTILITIES_API XMLValidator
+{
+ public:
+  /** @name Constructors */
+  //@{
+
+  /// Constructor for a new validator
+  explicit XMLValidator(const openstudio::path& schemaPath);
+
+  //@}
+  /** @name Getters */
+  //@{
+
+  openstudio::path schemaPath() const;
+
+  bool validate(const openstudio::path& xmlPath);
+
+  // Below functions are related to the last call to validate
+
+  bool isValid() const;
+  boost::optional<openstudio::path> xmlPath() const;
+
+  std::vector<LogMessage> errors() const;
+
+  std::vector<LogMessage> warnings() const;
+
+  /** Only applicable for Schematron validation **/
+  boost::optional<std::string> fullValidationReport() const;
+
+  //@}
+  /** @name Setters */
+  //@{
+
+  //@}
+  /** @name Operators */
+  //@{
+
+  //@}
+
+ private:
+  // TODO: I'm not sure we need that complexity... perhaps just use std::vector<LogMessage> m_logMessages;
+  REGISTER_LOGGER("openstudio.XMLValidator");
+  StringStreamLogSink m_logSink;
+  openstudio::path m_schemaPath;
+  boost::optional<openstudio::path> m_xmlPath;
+
+  XMLValidatorType m_validatorType;
+
+  bool xsdValidate() const;
+
+  // TODO: should we provide a facility to convert from schematron to XLST? or let user deal with that
+  bool xsltValidate() const;
+  mutable std::string m_fullValidationReport;
+
+  // reset the state of the XMLValidator between translations
+  void reset();
+
+  // A static Initializer, as a Meyers Singleton, to know when we need to init/cleanup the libxml2 & libxslt libs
+  // This lazily does the initialization, and clean up will be done completey at the end of the program
+  static detail::XMLInitializer& xmlInitializerInstance();
```

Comment:
In order to achieve the correct order, I use a Meyers singleton to lazily init a **static** XMLInitializer